cpp/server/ct-mirror.cc

```diff
@@ -0,0 +1,386 @@
+    //
+    // Put a sensible single-node config into FakeEtcd. For a real clustered
+    // log
+    // we'd expect a ClusterConfig already to be present within etcd as part of
+    // the provisioning of the log.
+    //
+    // TODO(alcutter): Note that we're currently broken wrt to restarting the
+    // log server when there's data in the log.  It's a temporary thing though,
+    // so fear ye not.
+    ct::ClusterConfig config;
+    config.set_minimum_serving_nodes(1);
+    config.set_minimum_serving_fraction(1);
+    LOG(INFO) << "Setting default single-node ClusterConfig:\n"
+              << config.DebugString();
+    server.consistent_store()->SetClusterConfig(config);
+
+    // Since we're a single node cluster, we'll settle that we're the
+    // master here, so that we can populate the initial STH
+    // (StrictConsistentStore won't allow us to do so unless we're master.)
+    server.election()->StartElection();
+    server.election()->WaitToBecomeMaster();
+  } else {
+    CHECK(!FLAGS_server.empty());
+  }
+
+  CHECK(!FLAGS_target_public_key.empty());
+  CHECK(!FLAGS_target_log_uri.empty());
+
+  const StatusOr<EVP_PKEY*> pubkey(ReadPublicKey(FLAGS_target_public_key));
+  CHECK(pubkey.ok()) << "Failed to read target log's public key file: "
+                     << pubkey.status();
+
+  ThreadPool pool(16);
+  SyncTask fetcher_task(&pool);
+
+  mutex queue_mutex;
+  map<int64_t, ct::SignedTreeHead> queue;
+
+  const function<void(const ct::SignedTreeHead&)> new_sth(
+      [&queue_mutex, &queue](const ct::SignedTreeHead& sth) {
+        lock_guard<mutex> lock(queue_mutex);
+        const auto it(queue.find(sth.tree_size()));
+        if (it != queue.end() && sth.timestamp() < it->second.timestamp()) {
+          LOG(WARNING) << "Received older STH:\nHad:\n"
+                       << it->second.DebugString() << "\nGot:\n"
+                       << sth.DebugString();
+          return;
+        }
+        queue.insert(make_pair(sth.tree_size(), sth));
+      });
+
+  const shared_ptr<RemotePeer> peer(make_shared<RemotePeer>(
+      unique_ptr<AsyncLogClient>(
+          new AsyncLogClient(&pool, &url_fetcher, FLAGS_target_log_uri)),
+      unique_ptr<LogVerifier>(
+          new LogVerifier(new LogSigVerifier(pubkey.ValueOrDie()),
+                          new MerkleVerifier(new Sha256Hasher))),
+      new_sth, fetcher_task.task()->AddChild(
+                   [](Task* task) { LOG(INFO) << "RemotePeer exited."; })));
+  const unique_ptr<ContinuousFetcher> fetcher(
+      ContinuousFetcher::New(event_base.get(), &pool, db));
+  fetcher->AddPeer("target", peer);
+
+  thread sth_updater(&STHUpdater, db, server.cluster_state_controller(),
+                     &queue_mutex, &queue,
+                     fetcher_task.task()->AddChild([](Task* task) {
+                       LOG(INFO) << "STHUpdater exited.";
+                     }));
+
+  server.Run();
+
+  fetcher_task.Cancel();
```

Comment:
You never give the actual `fetcher_task` to anything, you only make childs with it, so there's no one to "notice" that you cancelled it, so the `Wait()` below is going to be a loooong wait! :wink:

Make this `fetcher_task.Return();`, and you'll be on your way. :smiley: